src/control/lib/hardware/topology.go

```diff
@@ -262,3 +357,29 @@ func (t DeviceType) String() string {
 
 	return "unknown device type"
 }
+
+// TopologyFactory is a TopologyProvider that merges results from multiple other
+// TopologyProviders.
+type TopologyFactory struct {
+	providers []TopologyProvider
+}
+
+// GetTopology gets a merged master topology from all the topology providers.
+func (tf *TopologyFactory) GetTopology(ctx context.Context) (*Topology, error) {
+	newTopo := &Topology{}
+	for _, prov := range tf.providers {
+		topo, err := prov.GetTopology(ctx)
+		if err != nil {
+			return nil, err
+		}
+		newTopo.Merge(topo)
+	}
```

Comment:
I think this is fine for now, and perhaps always will be. It occurs to me to wonder whether or not we need to somehow decide which topologies should be grabbed first based on some kind of weighting mechanism, though. As it is, if there is an ordering dependency, it requires the person writing the code to understand it and make sure that the providers list is set up correctly.

Reply:
I added a weighting mechanism to the inputs for the NewTopologyFactory method.